File: 1513.number-of-substrings-with-only-1-s.js
/*
 * @lc app=leetcode id=1513 lang=javascript
 *
 * [1513] Number of Substrings With Only 1s
 */

// @lc code=start
/**
 * @param {string} s
 * @return {number}
 */
var numSub = function (s)
{
    const MOD = 10 ** 9 + 7;
    let numSubCount = 0;
    let currentSubCount = 0;

    for (const c of s)
    {
        if (c === '1')
        {
            currentSubCount++;
            numSubCount += currentSubCount;
        }
        else
        {
            currentSubCount = 0;
        }
    }

    return numSubCount % MOD;
};

// @lc code=end

File: src/2025/1513.number-of-substrings-with-only-1-s.cpp
/*
 * @lc app=leetcode id=1513 lang=cpp
 *
 * [1513] Number of Substrings With Only 1s
 */

#include <string_view>

// @lc code=start
class Solution {
 public:
  int numSub(std::string_view s) {
    static constexpr size_t kMod = 1e9 + 7;

    size_t left = 0;
    size_t right = 0;
    size_t subarray_count = 0;

    do {
      while (left < s.size() && s[left] == '0') {
        left++;
      }
      if (left == s.size()) {
        return subarray_count;
      }

      right = left;
      while (right < s.size() && s[right] == '1') {
        subarray_count += right - left + 1;
        subarray_count %= kMod;
        right++;
      }
      if (right == s.size()) {
        break;
      }
      left = right;
    } while (true);

    return subarray_count;
  }
};
// @lc code=end
